demo/neural-network-demo

jszuminski opened the pull request:
Added a neural network demo for the demo pages.

---
## Demo
Link to a JSFiddle: https://jsfiddle.net/BlackLabel/3whL2sfe/

## Things to consider
We might consider (in the future or now in this demo) adding a panel of inputs where the user could choose the number of layers and the number of nodes for each of the layers (but wouldn't this be an overkill in a simplistic demo that we want to release now?).

pawelfus commented on samples/highcharts/demo/neural-network/demo.js:
```diff
@@ -0,0 +1,101 @@
+            'to visualize a neural network.'
+    },
+    tooltip: {
+        formatter() {
```
Could you replace formatter with `format`? You can store extra info per series/point/axis.

pawelfus commented on samples/highcharts/demo/neural-network/demo.js:
```diff
@@ -0,0 +1,101 @@
+const layers = [1, 6, 6, 6, 2];
```
Add a comment that those variables can be modified to render different graphs.
Unless we want to go with a fixed dataset, leaving the decision up to @TorsteinHonsi

TorsteinHonsi replied:
+1 for the comment. I'd like to see some more comments in general for this demo.

pawelfus reviewed (changes requested):
Looks good!

Could you check why we see chart skew in mobile view? See: 
![image](https://github.com/highcharts/highcharts/assets/1453926/00b0acab-1088-4dfa-8c07-7aa3c54a9324)
 
I guess this might be related to computed yAxis extremes.- at least I would start there. Alternatively, can we reduce labels to short names (ReLU, tanh etc)?

jszuminski commented:
I've added the requested changes. This is an updated demo in JSFiddle: https://jsfiddle.net/47uaj3cs/

Also, I was thinking that maybe instead of `layers` and `activation` arrays of numbers, it would be more intuitive for the users to create an array of objects like this:
```
const layers = [{
  nodes: 6,
  activation: 'tanh'
}, {
  nodes: 6,
  activation: 'ReLU'
}, {
  nodes: 6,
  activation: 'ReLU'
}, {
  nodes: 2,
  activation: 'sigmoid'
}]
```

What do you think @pawelfus @TorsteinHonsi?

pawelfus commented:
Sounds good!

jszuminski commented:
Here's an updated JSFiddle: https://jsfiddle.net/4p2bLkhw/

Ready for rereview @pawelfus

pawelfus reviewed (approved):
Thanks!

pawelfus commented:
Adding Marita for a11y review

jszuminski commented:
Most up to date demo: https://jsfiddle.net/ewcxdh5b/

jszuminski commented on samples/highcharts/demo/neural-network/demo.js:
```diff
@@ -0,0 +1,137 @@
+    label: 'Output Layer (sigmoid)'
+}];
+
+function generateData() {
```
I've tried my best to describe it with a few comments, though not sure if that's fully clear.

TorsteinHonsi replied:
Looks good 👍

jszuminski commented:
@TorsteinHonsi most up to date demo: https://jsfiddle.net/BlackLabel/2a1y5zhe/

I've removed the `colors` array and added it to `plotOptions.line.color`, but I've also reduced the opacity of the lines.

In my opinion, it looks better now. What do you think?

TorsteinHonsi reviewed (approved):
Looks good, thank you!

jszuminski commented:
As discussed with @marvin19, we are removing `point.descriptionFormat` temporarily as there are a few problems related to it independent to this pull request.

TorsteinHonsi reviewed (changes requested):
Thanks!

1. The `descriptionFormat` works for me like this
```js
    accessibility: {
        typeDescription: 'Neural network chart',
        point: {
            descriptionFormat:
                'node on {series.xAxis.options.custom.layers.(x).label}'
        }
    },
```
Is that okay or are there other problems?


2. The files of the `vendor` folder are added in this PR. Those should be removed.
